aws/resource_aws_budget.go

```diff
@@ -0,0 +1,303 @@
+		budgetName = id
+
+	} else if v, ok := d.GetOk("name"); ok {
+		budgetName = v.(string)
+
+	} else if v, ok := d.GetOk("name_prefix"); ok {
+		budgetName = resource.PrefixedUniqueId(v.(string))
+
+	} else {
+		budgetName = resource.UniqueId()
+	}
+
+	budgetType := d.Get("budget_type").(string)
+	budgetLimitAmount := d.Get("limit_amount").(string)
+	budgetLimitUnit := d.Get("limit_unit").(string)
+	budgetIncludeCredit := d.Get("include_credit").(bool)
+	budgetIncludeOtherSubscription := d.Get("include_other_subscription").(bool)
+	budgetIncludeRecurring := d.Get("include_recurring").(bool)
+	budgetIncludeRefund := d.Get("include_refund").(bool)
+	budgetIncludeSubscription := d.Get("include_subscription").(bool)
+	budgetIncludeSupport := d.Get("include_support").(bool)
+	budgetIncludeTax := d.Get("include_tax").(bool)
+	budgetIncludeUpfront := d.Get("include_upfront").(bool)
+	budgetUseBlended := d.Get("use_blended").(bool)
+	budgetTimeUnit := d.Get("time_unit").(string)
+	budgetCostFilters := make(map[string][]*string)
+	for k, v := range d.Get("cost_filters").(map[string]interface{}) {
+		filterValue := v.(string)
+		budgetCostFilters[k] = append(budgetCostFilters[k], &filterValue)
+	}
+
+	budgetTimePeriodStart, err := time.Parse("2006-01-02_15:04", d.Get("time_period_start").(string))
+	if err != nil {
+		return nil, fmt.Errorf("failure parsing time: %v", err)
+	}
+
+	budgetTimePeriodEnd, err := time.Parse("2006-01-02_15:04", d.Get("time_period_end").(string))
+	if err != nil {
+		return nil, fmt.Errorf("failure parsing time: %v", err)
+	}
+
+	budget := new(budgets.Budget)
+	budget.SetBudgetName(budgetName)
+	budget.SetBudgetType(budgetType)
+	budget.SetBudgetLimit(&budgets.Spend{
+		Amount: &budgetLimitAmount,
+		Unit:   &budgetLimitUnit,
+	})
+	budget.SetCostTypes(&budgets.CostTypes{
+		IncludeCredit:            &budgetIncludeCredit,
+		IncludeOtherSubscription: &budgetIncludeOtherSubscription,
+		IncludeRecurring:         &budgetIncludeRecurring,
+		IncludeRefund:            &budgetIncludeRefund,
+		IncludeSubscription:      &budgetIncludeSubscription,
+		IncludeSupport:           &budgetIncludeSupport,
+		IncludeTax:               &budgetIncludeTax,
+		IncludeUpfront:           &budgetIncludeUpfront,
+		UseBlended:               &budgetUseBlended,
+	})
+	budget.SetTimePeriod(&budgets.TimePeriod{
+		End:   &budgetTimePeriodEnd,
+		Start: &budgetTimePeriodStart,
+	})
+	budget.SetTimeUnit(budgetTimeUnit)
+	budget.SetCostFilters(budgetCostFilters)
+	return budget, nil
+}
+
+func budgetExists(budgetName string, meta interface{}) bool {
+	_, err := describeBudget(budgetName, meta)
+	if isBudgetNotFoundException(err) {
+		return false
+	}
+
+	return true
+}
+
+func isBudgetNotFoundException(err error) bool {
+	if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == budgets.ErrCodeNotFoundException {
+		return true
+	}
+
+	return false
+}
+
+func describeBudget(budgetName string, meta interface{}) (*budgets.DescribeBudgetOutput, error) {
```

Comment:
This function seems extraneous as well. Its not really saving any complexity and is just introducing additional redirection.